### PR TITLE
2.32.0 prometheus cw

### PR DESCRIPTION
**Pull request by matthewfala**
Example Updates
1. Add cloudwatch_logs note on using multiple workers for higher throughput in version 2.32.0 and beyond
2. Add Amazon Managed Prometheus comment to explain how to send metrics to your own personal amazon prometheus workspace, and grafana setup link to aws docs

[Related to the 2.32.0 update](https://github.com/aws/aws-for-fluent-bit/pull/733)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/fluent-bit/amazon-managed-service-for-prometheus/permissions.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{
+    "Version": "2012-10-17",
```
Comment (matthewfala) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/permissions.json:2]: Strict iam policies tested

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+## FireLens Example: Prometheus forward metrics - using the Fluent Bit image with init tag
```
Comment (PettitWesley) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: I think this needs a link in the main readme

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+This example shows how to forward metrics with Fluent Bit's prometheus remote write output plugin to an Amazon Managed Service for Prometheus workspace.
+
+### Step 1: Create config file locally
+
+**fluent-bit.conf**
+
+```
+# FireLens Example: Prometheus forward metrics - using the Fluent Bit image with init tag
+
+# Scape node metrics every 20 seconds (use any metrics input plugin)
```
Comment (PettitWesley) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Scrape

Comment (PettitWesley) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: > (use any metrics input plugin)

What does this mean? This sentence confuses me.

Comment (PettitWesley) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Link input docs

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    Uri /workspaces/< my-amp-workspace-id >/api/v1/remote_write
+    AWS_Auth On
+    AWS_region us-west-2
+    Tls On
+    Tls.verify On
+    add_label app my-ecs-app
+    add_label color blue
+```
+
+**Note:** you can find this config file in the `config-files` directory of this example, please modify according to the actual situation to match your needs.
```
Comment (PettitWesley) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: > actual situation

your resource values? can this language be more technical sounding?

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+### Step 2: Create an Amazon Managed Service for Prometheus Workspace
+
+* create an Amazon Managed Service for Prometheus workspace `my-prometheus` to store metrics
+* update above config file's `prometheus_remote_write` `uri` to the created workspace's remote write url endpoint
```
Comment (PettitWesley) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Capitalize start of sentences?

Comment (PettitWesley) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Add doc links for creating workspace?

Comment (PettitWesley) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Should this be step 1, create your workspace then create your config based on it?

File: examples/fluent-bit/cloudwatchlogs-emf/emf-over-tcp/extra.conf [hunk 6]
```diff
@@ -24,3 +24,6 @@
     log_stream_prefix   from-fluent-bit-
     auto_create_group   true
     log_format          json/emf
+    # from aws-for-fluent-bit v2.32.0 and on, to support higher throughput logging,
+    # set workers to a high value such as 5 or the number of cores on your host
```
Comment (PettitWesley) [inline examples/fluent-bit/cloudwatchlogs-emf/emf-over-tcp/extra.conf]: I'd rephrase this so rather than being a command its like, "consider configuring more than 1 worker, up to the number of cores on your host"

and I'd also explicitly say only 2.32+ supports multi worker

Comment (PettitWesley) [reply examples/fluent-bit/cloudwatchlogs-emf/emf-over-tcp/extra.conf]: phrase it that way

File: examples/fluent-bit/cloudwatchlogs/README.md [hunk 7]
```diff
@@ -7,12 +7,16 @@ There are two Fluent Bit output plugins for sending to Amazon CloudWatch Logs:
...
+For more on the AWS Go outputs vs AWS C outputs, check out the [FAQ entry in our debugging guide](https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#aws-go-plugins-vs-aws-core-c-plugins).
 
 #### Recommended cloudwatch_logs configuration options
 
 To minimize the possibility of log loss when sending to CloudWatch, consider using the recommended configuration outlined in the [CloudWatch Recommendations](https://github.com/aws/aws-for-fluent-bit/issues/340) issue.
 
+#### High throughput logging to CloudWatch Logs via multiple workers
+
+As of AWS For Fluent Bit `2.32.0`, the `cloudwatch_out` plugin supports high througpht logging via multiple workers. Set the `workers` option to a high value such as `5`. Fluent Bit will spin up the specified number of workers as threads which will take advantage of concurrency. For best performance, the number of workers should be at least the number of cores on the host. AWS For Fluent Bit `2.31.12` and prior does not support multiple workers and should should only use `0` and `1` workers.
```
Comment (PettitWesley) [inline examples/fluent-bit/cloudwatchlogs/README.md]: > `cloudwatch_out` plugin

?

Comment (PettitWesley) [reply examples/fluent-bit/cloudwatchlogs/README.md]: througpht

Comment (PettitWesley) [reply examples/fluent-bit/cloudwatchlogs/README.md]: > For best performance, the number of workers should be at least the number of cores on the host.

What's this based on? I wouldn't say this. I don't think we can back this up. 

> should should only use `0` and `1` workers.

Do we ever want to recommend 0 workers in ECS?

File: examples/fluent-bit/oomkill-prevention/firelens-full-filesystem-example/fluent-bit-image/extra.conf [hunk 8]
```diff
@@ -49,4 +49,7 @@
     auto_create_group true
     retry_limit 2
     # limit the filesystem storage that this output can use
-    storage.total_limit_size 1G
\ No newline at end of file
+    storage.total_limit_size 1G
+    # from aws-for-fluent-bit v2.32.0 and on, to support higher throughput logging,
+    # set workers to a high value such as 5 or the number of cores on your host
```
Comment (PettitWesley) [inline examples/fluent-bit/oomkill-prevention/firelens-full-filesystem-example/fluent-bit-image/extra.conf]: Can you please replace all occurrences of "high value" with a different phrasing?

File: examples/fluent-bit/cloudwatchlogs/README.md (hunk 7 above)
Comment (matthewfala) [reply examples/fluent-bit/cloudwatchlogs/README.md]: I think 1 worker is a good option.

Threads are usually advantageous when you have multiple cores to take advantage of parallel processing. Nothing to back it up in the code.

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 2 above)
Comment (matthewfala) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Added

File: examples/fluent-bit/oomkill-prevention/firelens-full-filesystem-example/fluent-bit-image/extra.conf (hunk 8 above)
Comment (matthewfala) [reply examples/fluent-bit/oomkill-prevention/firelens-full-filesystem-example/fluent-bit-image/extra.conf]: Changed wording to recommended wording

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 5 above)
Comment (matthewfala) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Added link, and changed

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 4 above)
Comment (matthewfala) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Updated wording

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 3 above)
Comment (matthewfala) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Updated

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 2 above)
Comment (PettitWesley) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Also:

> Prometheus forward metrics

This confuses me and makes me think of forward input. I'd change to: forward metrics to Prometheus - using...

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    Uri /workspaces/< my-amp-workspace-id >/api/v1/remote_write
+    AWS_Auth On
+    AWS_region us-west-2
+    Tls On
+    Tls.verify On
+    add_label app my-ecs-app
+    add_label color blue
+```
+
+**Note:** You can find this config file in the `config-files` directory of this example. Please modify the inputs according to the metrics you desire to collect, and resource values to match your application's setup.
```
Comment (PettitWesley) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: > modify the inputs

modify the `node_exporter_metrics` input config options according...

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+### Step 3: Upload config file to S3
+
+* Create the S3 bucket `your-bucket` to store config files
+* Upload above config file to this bucket
+
+### Step 4: Create the ECS Task
+
+* Create the ECS Task using provided `task-definition.json`, which using the Fluent Bit image with init tag
+* Change the `taskRoleArn` and `executionRoleArn` to your own role ARN
+* Change the `environment` part in the task definition FireLens configuration, copy the ARN of config files and paste it as environment variable's value. The name of environment variable requires to use the prefix `aws_fluent_bit_init_s3_`
```
Comment (PettitWesley) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: > Change the `environment` part in the task definition FireLens configuration,

Edit the `environment` section container definition for the FireLens container; copy...

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 9 above)
Comment (PettitWesley) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Can you make this more specific/descriptive please?

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 10 above)
Comment (PettitWesley) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Can you make this more specific/descriptive please?

File: examples/fluent-bit/cloudwatchlogs/README.md [hunk 11]
```diff
@@ -7,12 +7,16 @@ There are two Fluent Bit output plugins for sending to Amazon CloudWatch Logs:
...
+For more on the AWS Go outputs vs AWS C outputs, check out the [FAQ entry in our debugging guide](https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#aws-go-plugins-vs-aws-core-c-plugins).
 
 #### Recommended cloudwatch_logs configuration options
 
 To minimize the possibility of log loss when sending to CloudWatch, consider using the recommended configuration outlined in the [CloudWatch Recommendations](https://github.com/aws/aws-for-fluent-bit/issues/340) issue.
 
+#### High throughput logging to CloudWatch Logs via multiple workers
+
+As of AWS For Fluent Bit `2.32.0`, the `cloudwatch_logs` plugin supports high throughput logging via multiple workers. Set the `workers` option to a high value such as `5`. Fluent Bit will spin up the specified number of workers as threads which will take advantage of concurrency. To optimize for high throughput logging, consider setting the number of workers the number of cores on the host. AWS For Fluent Bit `2.31.12` and prior does not support multiple workers, and `workers` should be set to `1`.
```
Comment (PettitWesley) [inline examples/fluent-bit/cloudwatchlogs/README.md]: > Set the `workers` option to a high value such as `5`. Fluent Bit will spin up the specified number of workers as threads which will take advantage of concurrency.

Please remove "high value" phrasing, and I'd say something like "Fluent Bit will allocate the output the specified number of worker threads, which can upload logs concurrently"

**Review by PettitWesley — CHANGES_REQUESTED** [review]

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
+## FireLens Example: Forward metrics to Prometheus using the Fluent Bit image with init tag
```
Comment (matthewfala) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Updated wording here.

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    Uri /workspaces/< my-amp-workspace-id >/api/v1/remote_write
+    AWS_Auth On
+    AWS_region us-west-2
+    Tls On
+    Tls.verify On
+    add_label app my-ecs-app
+    add_label color blue
+```
+
+**Note:** You can find this config file in the `config-files` directory of this example. Please modify the `node_exporter_metrics` input configuration options according to the metrics you desire to collect, along with the `prometheus_remote_write` output configuration options for `Host` and `Uri` to match your AMP setup.
```
Comment (matthewfala) [inline examples/fluent-bit/amazon-managed-service-for-prometheus/README.md:40]: Updated wording here, and in Step 4

File: examples/fluent-bit/cloudwatchlogs/README.md [hunk 14]
```diff
@@ -7,12 +7,16 @@ There are two Fluent Bit output plugins for sending to Amazon CloudWatch Logs:
...
+For more on the AWS Go outputs vs AWS C outputs, check out the [FAQ entry in our debugging guide](https://github.com/aws/aws-for-fluent-bit/blob/mainline/troubleshooting/debugging.md#aws-go-plugins-vs-aws-core-c-plugins).
 
 #### Recommended cloudwatch_logs configuration options
 
 To minimize the possibility of log loss when sending to CloudWatch, consider using the recommended configuration outlined in the [CloudWatch Recommendations](https://github.com/aws/aws-for-fluent-bit/issues/340) issue.
 
+#### High throughput logging to CloudWatch Logs via multiple workers
+
+As of AWS For Fluent Bit `2.32.0`, the `cloudwatch_logs` plugin supports high throughput logging via multiple workers. Set the `workers` option to an integer value, such as `5`, indicating the number of worker threads dedicated to processing output data to CloudWatch Logs concurrently. To optimize for high throughput logging, consider setting the number of workers the number of cores on the host. AWS For Fluent Bit `2.31.12` and prior does not support multiple workers, and `workers` should be set to `1`.
```
Comment (matthewfala) [inline examples/fluent-bit/cloudwatchlogs/README.md:18]: Updated wording on thread here.

File: examples/fluent-bit/amazon-managed-service-for-prometheus/README.md (hunk 2 above)
Comment (matthewfala) [reply examples/fluent-bit/amazon-managed-service-for-prometheus/README.md]: Sure! Updated.

File: examples/fluent-bit/cloudwatchlogs-emf/emf-over-tcp/extra.conf (hunk 6 above)
Comment (matthewfala) [reply examples/fluent-bit/cloudwatchlogs-emf/emf-over-tcp/extra.conf]: Sure! Updated.

**Comment by matthewfala** [comment]
Updated. Would you please take another look?

**Review by weijianz-opc — APPROVED** [review]
lgtm

**Review by PettitWesley — APPROVED** [review]
